test/index.test.ts

```diff
@@ -14,6 +36,81 @@ interface MockUser {
   email: string;
 }
 
+type ReturnTypeFromStage<
+  T extends (context: any) => any,
+  TStage extends AllStageKeys
+> = TStage extends AllAsyncStageKeys
+  ? PromiseResolveOrSync<ReturnType<T>>
+  : ReturnType<T>;
+
+async function HTPipeTest<
+  TPipe extends OptionallyHasInitPreContext<any, any> &
+    OptionallyHasSanitizeParams<any, any> &
+    OptionallyHasSanitizeBody<any, any> &
+    MightHavePreAuthorize<any, any> &
+    OptionallyHasAttachData<any, any> &
+    MightHaveFinalAuthorize<any, any> &
+    OptionallyHasDoWork<any, any> &
+    MightHaveRespond<any, any> &
+    MightHaveSanitizeResponse<any, any>,
+  TPipeIn extends TPipeInExpected,
+  TPipeOut extends TPipeOutExpected,
+  TStage extends AllStageKeys,
+  TLifecycleStage extends TPipe extends HasInitPreContext<any, any>
+    ? TPipe['initPreContext']
+    : TPipe extends HasSanitizeParams<any, any>
+    ? TPipe['sanitizeParams']
+    : TPipe extends HasSanitizeBody<any, any>
+    ? TPipe['sanitizeBody']
+    : TPipe extends HasPreAuthorize<any, any>
+    ? TPipe['preAuthorize']
+    : TPipe extends HasAttachData<any, any>
+    ? TPipe['attachData']
+    : TPipe extends HasFinalAuthorize<any, any>
+    ? TPipe['finalAuthorize']
+    : TPipe extends HasDoWork<any, any>
+    ? TPipe['doWork']
+    : TPipe extends HasRespond<any, any>
+    ? TPipe['respond']
+    : TPipe extends HasSanitizeResponse<any, any>
+    ? TPipe['sanitizeResponse']
+    : never,
+  TValid extends TPipeInExpected extends Parameters<TLifecycleStage>[0]
+    ? Parameters<TLifecycleStage>[0] extends TPipeInExpected
+      ? TPipeOutExpected extends ReturnTypeFromStage<TLifecycleStage, TStage>
+        ? ReturnTypeFromStage<TLifecycleStage, TStage> extends TPipeOutExpected
+          ? true
+          : never
+        : never
+      : never
+    : never,
+  TPipeInExpected = TPipeIn,
+  TPipeOutExpected = TPipeOut
+>(
+  pipe: TPipe,
+  lifecycleStage: TStage,
+  pipeIn: TPipeIn,
+  pipeOut: TPipeOut,
+  valid: TValid,
+  pipeInExpected: TPipeInExpected = pipeIn,
+  pipeOutExpected: TPipeOutExpected = pipeOut
+) {
+  const pipedLifecycleStage = pipe[lifecycleStage];
+
```

Comment:
Can you explain this if/else?  I don't quite get it - isn't it a tautology that the else branch will always fail?  Or is there some truthiness edge case I'm not thinking of with the cases where we want things to be empty?

Reply:
- [ ] explain

to be a little clearer, on the surface it looks like this code can/should just be:
```
expect(pipedLifecycleStage).to.not.be.empty;
if(pipedLifecycleStage) {
...
}
```

Reply:
first of all you right and that else block is no needed
but if conditions is needed because without that condition i can't use the lifecycle() because it can be undefined